torch/distributed/fsdp/fully_sharded_data_parallel.py

```diff
@@ -461,6 +465,31 @@ def __init__(
             if param not in ignored_params and not isinstance(param, FlatParameter):
                 params.append(param)
 
+        # At this point we are wrapping an atomic FSDP unit. Check module shapes
+        # and broadcast module from rank 0 to ensure everything starts off with
+        # the same parameters.
+        # Hack: move to compute device and back if it is CPU
+        prev_device = next(module.parameters()).device
+        was_cpu = (prev_device == torch.device("cpu"))
+        if was_cpu:
+            warnings.warn(
+                f"Module is input on CPU, we are moving to to {self.compute_device} to perform parameter verification and will move it back after. In the future please specify device_ids which will explicitly tell FSDP to place the module on a particular device."
+            )
+            module = module.to(self.compute_device)
+        _verify_param_shape_across_processes(
+            process_group=self.process_group,
+            tensors=params,  # TODO: FSDP + DDP to verify buffers as well
+        )
+        _sync_params_and_buffers(
+            module=module,
+            process_group=self.process_group,
+            # Same bucket size as used in DDP.
+            broadcast_bucket_size=int(250 * 1024 * 1024),
+            rank=0,
+            params_and_buffers_to_ignore=ignored_params,
+        )
+        if was_cpu:
+            module = module.to(prev_device)
```

Comment:
Could we move this to be after "self._shard_parameters()", so that we can flatten and shard parameters on GPU, this will greatly improve model init time when users pass a CPU model without device_id specified?

Reply:
This is taken care of if users pass in `device_id`s already. Also this PR and `was_cpu` logic is mostly intended for discussion, as this does not seem like a great way to support CPU use cases?

Reply:
that sounds good we only support GPU sharding when users pass a device id. although we can support GPU sharding and move sharded module from GPU to CPU even users do not pass a device id? but probably it is not a good user pattern...

Reply:
Sure, per our discussion we can offer basic support for this and log a warning saying device_ids should be used. Will add warning in the next diff for device_Ids.

For this PR, if CPU, we'll move to compute_device (in the future compute_device == device_id should be an invariant), and run sharding, flattening as well on GPU before moving the module back to CPU.

Reply:
We also call `del module` after sharding and before flattening, I think given this, if original model was on CPU, we can simply move it back similar to how we offload the param to CPU in ctor.